crates/ra_hir/src/code_model.rs

```diff
@@ -368,6 +375,14 @@ impl Struct {
     }
 }
 
+impl HasVisibility for Struct {
+    fn visibility(&self, db: &dyn HirDatabase) -> Visibility {
+        let struct_data = db.struct_data(self.id);
+        let visibility = &struct_data.visibility;
+        visibility.resolve(db.upcast(), &self.id.resolver(db.upcast()))
```

Comment:
Maybe @matklad can clarify. Do we by-design always accept `HirDatabase` even if we only need `DefDatabase` in `ra_hir`?

Reply:
In this case, we need `HirDatabase`, it is because the `trait` is called by same crate which should only exposed `HirDatabase` interface. I doubt that it is salsa limitation such that it has to be using an inheritance structure instead of composition to represent the hierarchy.   But feel free to correct me if i miss understand.